reps/2023-02-01-wasm-on-ray.md

```diff
@@ -0,0 +1,253 @@
+
+Furthermore, it is also possible that we enable transparent Ray support in WASM. As a result, many legacy applications can be easily migrated to Ray and run in a distributed manner and Ray layer can be transparent to the application. This will be a huge benefit for Ray adoption in the generic serverless computing environment.
+
+In our internal FaaS environment, we have WASM function support. These functions are compiled from different programming languages such as Javascript, Rust, Go, C/C++, etc. Compared with normal container based FaaS solutions such as Knative, WASM based functions can be loaded with a shorter period of time while still providing the same or even better secure running environments. So far, WASM functions are still far from perfect. There are few pain points for WASM functions.
+
+1. If different functions want to communicate with each other, they need use either RPC, message queue or KV store. However, these solutions are not convenient for the developers. The FaaS function developers need a simpler communication mechanism between different functions and run tasks in parallel.
+2. WASM function is executed in a single thread. Although multi-threading support is on the way, it is still not mature enough. To run tasks in a single FaaS function in parallel, running WASM with Ray API support is a good solution. Besides, running a WASM function distributedly in Ray has the FT support out of the box.
+3. Our FaaS environment has a large number of edge servers. The resources on each of the servers are limited. It is not possible to provide more resources for a single function than the resources on the server. In Ray, all the resources in the Ray cluster is treated as a single resource pool. If we can support Ray in our FaaS environment, we can run WASM functions with larger resource requirements.
+
+As a result, Having Ray support in our FaaS WASM environment will be a huge benefit for us and it is a very important feature for us.
+
+One use case of WASM on Ray is the ML inference tasks. In our enterprise environment, we need to support running ML inference tasks in Ray and make these ML inference tasks share the Ray cluster resources. Compared to the current Ray Serve ML inference solution, which needs to keep the ML inference tasks running in the Ray cluster memory, the WASM ML inference solution can reduce the resource consumption of the Ray cluster. Currently, WASM runtime such as WasmEdge already support running ML inference tasks through WASI-NN support.
```

Comment:
I think for Ray Serve ML inference, the ML also runs in the memory. Could you give more detail about the difference between going with Ray Serve and WASI-NN ?

Reply:
wasi-nn is a low-level ML computing API. https://github.com/WebAssembly/wasi-nn/blob/main/wasi-nn.wit.md such as loading a model, setting input, computing with the model, and getting the output. In contrast, Ray Serve is a high-level Python API for the model serving. In order to do something similar in WASM, we need to port some of the Ray Serve from the Python worker to the WASM worker.